File: Patterns/patterns.py
"""
Pattern Printing Programs
Common patterns asked in junior developer coding interviews
"""


def print_square_pattern(n):
    """
    Pattern 1: Square Pattern
    * * * * *
    * * * * *
    * * * * *
    * * * * *
    * * * * *
    """
    print(f"\n1. Square Pattern ({n}x{n}):")
    print("-" * 40)
    for i in range(n):
        for j in range(n):
            print("*", end=" ")
        print()


def print_right_triangle(n):
    """
    Pattern 2: Right-Angled Triangle
    *
    * *
    * * *
    * * * *
    * * * * *
    """
    print(f"\n2. Right-Angled Triangle ({n} rows):")
    print("-" * 40)
    for i in range(1, n + 1):
        for j in range(i):
            print("*", end=" ")
        print()


def print_inverted_right_triangle(n):
    """
    Pattern 3: Inverted Right-Angled Triangle
    * * * * *
    * * * *
    * * *
    * *
    *
    """
    print(f"\n3. Inverted Right-Angled Triangle ({n} rows):")
    print("-" * 40)
    for i in range(n, 0, -1):
        for j in range(i):
            print("*", end=" ")
        print()


def print_pyramid(n):
    """
    Pattern 4: Pyramid
        *
       * *
      * * *
     * * * *
    * * * * *
    """
    print(f"\n4. Pyramid ({n} rows):")
    print("-" * 40)
    for i in range(1, n + 1):
        # Print spaces
        for j in range(n - i):
            print(" ", end=" ")
        # Print stars
        for k in range(i):
            print("*", end=" ")
        print()


def print_inverted_pyramid(n):
    """
    Pattern 5: Inverted Pyramid
    * * * * *
     * * * *
      * * *
       * *
        *
    """
    print(f"\n5. Inverted Pyramid ({n} rows):")
    print("-" * 40)
    for i in range(n, 0, -1):
        # Print spaces
        for j in range(n - i):
            print(" ", end=" ")
        # Print stars
        for k in range(i):
            print("*", end=" ")
        print()


def print_diamond(n):
    """
    Pattern 6: Diamond
        *
       * *
      * * *
     * * * *
    * * * * *
     * * * *
      * * *
       * *
        *
    """
    print(f"\n6. Diamond ({n} rows):")
    print("-" * 40)
    # Upper half (pyramid)
    for i in range(1, n + 1):
        for j in range(n - i):
            print(" ", end=" ")
        for k in range(i):
            print("*", end=" ")
        print()
    # Lower half (inverted pyramid)
    for i in range(n - 1, 0, -1):
        for j in range(n - i):
            print(" ", end=" ")
        for k in range(i):
            print("*", end=" ")
        print()


def print_hollow_square(n):
    """
    Pattern 7: Hollow Square
    * * * * *
    *       *
    *       *
    *       *
    * * * * *
    """
    print(f"\n7. Hollow Square ({n}x{n}):")
    print("-" * 40)
    for i in range(n):
        for j in range(n):
            if i == 0 or i == n - 1 or j == 0 or j == n - 1:
                print("*", end=" ")
            else:
                print(" ", end=" ")
        print()


def print_number_pattern(n):
    """
    Pattern 8: Number Pattern
    1
    1 2
    1 2 3
    1 2 3 4
    1 2 3 4 5
    """
    print(f"\n8. Number Pattern ({n} rows):")
    print("-" * 40)
    for i in range(1, n + 1):
        for j in range(1, i + 1):
            print(j, end=" ")
        print()


def print_number_pyramid(n):
    """
    Pattern 9: Number Pyramid
        1
       2 2
      3 3 3
     4 4 4 4
    5 5 5 5 5
    """
    print(f"\n9. Number Pyramid ({n} rows):")
    print("-" * 40)
    for i in range(1, n + 1):
        for j in range(n - i):
            print(" ", end=" ")
        for k in range(i):
            print(i, end=" ")
        print()


def print_floyd_triangle(n):
    """
    Pattern 10: Floyd's Triangle
    1
    2 3
    4 5 6
    7 8 9 10
    11 12 13 14 15
    """
    print(f"\n10. Floyd's Triangle ({n} rows):")
    print("-" * 40)
    num = 1
    for i in range(1, n + 1):
        for j in range(i):
            print(num, end=" ")
            num += 1
        print()


def print_pascal_triangle(n):
    """
    Pattern 11: Pascal's Triangle
         1
        1 1
       1 2 1
      1 3 3 1
     1 4 6 4 1
    """
    print(f"\n11. Pascal's Triangle ({n} rows):")
    print("-" * 40)
    for i in range(n):
        # Print spaces
        for j in range(n - i - 1):
            print(" ", end=" ")
        # Calculate and print values
        num = 1
        for j in range(i + 1):
            print(num, end=" ")
            num = num * (i - j) // (j + 1)
        print()


def print_alphabet_pattern(n):
    """
    Pattern 12: Alphabet Pattern
    A
    A B
    A B C
    A B C D
    A B C D E
    """
    print(f"\n12. Alphabet Pattern ({n} rows):")
    print("-" * 40)
    for i in range(n):
        for j in range(i + 1):
            print(chr(65 + j), end=" ")
        print()


# Test all patterns
if __name__ == "__main__":
    print("=" * 60)
    print("PATTERN PRINTING PROGRAMS FOR JUNIOR DEVELOPERS")
    print("=" * 60)
    
    n = 5  # Size of patterns
    
    print_square_pattern(n)
    print_right_triangle(n)
    print_inverted_right_triangle(n)
    print_pyramid(n)
    print_inverted_pyramid(n)
    print_diamond(n)
    print_hollow_square(n)
    print_number_pattern(n)
    print_number_pyramid(n)
    print_floyd_triangle(n)
    print_pascal_triangle(n)
    print_alphabet_pattern(n)
    
    print("\n" + "=" * 60)
    print("All patterns generated successfully!")
    print("=" * 60)

File: Patterns/Patterns.java
/**
 * Pattern Printing Programs
 * Common patterns asked in junior developer coding interviews
 */

public class Patterns {

    /**
     * Pattern 1: Square Pattern
     */
    public static void printSquarePattern(int n) {
        System.out.println("\n1. Square Pattern (" + n + "x" + n + "):");
        System.out.println("----------------------------------------");
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                System.out.print("* ");
            }
            System.out.println();
        }
    }

    /**
     * Pattern 2: Right-Angled Triangle
     */
    public static void printRightTriangle(int n) {
        System.out.println("\n2. Right-Angled Triangle (" + n + " rows):");
        System.out.println("----------------------------------------");
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                System.out.print("* ");
            }
            System.out.println();
        }
    }

    /**
     * Pattern 3: Inverted Right-Angled Triangle
     */
    public static void printInvertedRightTriangle(int n) {
        System.out.println("\n3. Inverted Right-Angled Triangle (" + n + " rows):");
        System.out.println("----------------------------------------");
        for (int i = n; i > 0; i--) {
            for (int j = 0; j < i; j++) {
                System.out.print("* ");
            }
            System.out.println();
        }
    }

    /**
     * Pattern 4: Pyramid
     */
    public static void printPyramid(int n) {
        System.out.println("\n4. Pyramid (" + n + " rows):");
        System.out.println("----------------------------------------");
        for (int i = 1; i <= n; i++) {
            // Print spaces
            for (int j = 0; j < n - i; j++) {
                System.out.print("  ");
            }
            // Print stars
            for (int k = 0; k < i; k++) {
                System.out.print("* ");
            }
            System.out.println();
        }
    }

    /**
     * Pattern 5: Inverted Pyramid
     */
    public static void printInvertedPyramid(int n) {
        System.out.println("\n5. Inverted Pyramid (" + n + " rows):");
        System.out.println("----------------------------------------");
        for (int i = n; i > 0; i--) {
            // Print spaces
            for (int j = 0; j < n - i; j++) {
                System.out.print("  ");
            }
            // Print stars
            for (int k = 0; k < i; k++) {
                System.out.print("* ");
            }
            System.out.println();
        }
    }

    /**
     * Pattern 6: Diamond
     */
    public static void printDiamond(int n) {
        System.out.println("\n6. Diamond (" + n + " rows):");
        System.out.println("----------------------------------------");
        // Upper half (pyramid)
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < n - i; j++) {
                System.out.print("  ");
            }
            for (int k = 0; k < i; k++) {
                System.out.print("* ");
            }
            System.out.println();
        }
        // Lower half (inverted pyramid)
        for (int i = n - 1; i > 0; i--) {
            for (int j = 0; j < n - i; j++) {
                System.out.print("  ");
            }
            for (int k = 0; k < i; k++) {
                System.out.print("* ");
            }
            System.out.println();
        }
    }

    /**
     * Pattern 7: Hollow Square
     */
    public static void printHollowSquare(int n) {
        System.out.println("\n7. Hollow Square (" + n + "x" + n + "):");
        System.out.println("----------------------------------------");
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 || i == n - 1 || j == 0 || j == n - 1) {
                    System.out.print("* ");
                } else {
                    System.out.print("  ");
                }
            }
            System.out.println();
        }
    }

    /**
     * Pattern 8: Number Pattern
     */
    public static void printNumberPattern(int n) {
        System.out.println("\n8. Number Pattern (" + n + " rows):");
        System.out.println("----------------------------------------");
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= i; j++) {
                System.out.print(j + " ");
            }
            System.out.println();
        }
    }

    /**
     * Pattern 9: Number Pyramid
     */
    public static void printNumberPyramid(int n) {
        System.out.println("\n9. Number Pyramid (" + n + " rows):");
        System.out.println("----------------------------------------");
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < n - i; j++) {
                System.out.print("  ");
            }
            for (int k = 0; k < i; k++) {
                System.out.print(i + " ");
            }
            System.out.println();
        }
    }

    /**
     * Pattern 10: Floyd's Triangle
     */
    public static void printFloydTriangle(int n) {
        System.out.println("\n10. Floyd's Triangle (" + n + " rows):");
        System.out.println("----------------------------------------");
        int num = 1;
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                System.out.print(num + " ");
                num++;
            }
            System.out.println();
        }
    }

    /**
     * Pattern 11: Pascal's Triangle
     */
    public static void printPascalTriangle(int n) {
        System.out.println("\n11. Pascal's Triangle (" + n + " rows):");
        System.out.println("----------------------------------------");
        for (int i = 0; i < n; i++) {
            // Print spaces
            for (int j = 0; j < n - i - 1; j++) {
                System.out.print("  ");
            }
            // Calculate and print values
            int num = 1;
            for (int j = 0; j <= i; j++) {
                System.out.print(num + " ");
                num = num * (i - j) / (j + 1);
            }
            System.out.println();
        }
    }

    /**
     * Pattern 12: Alphabet Pattern
     */
    public static void printAlphabetPattern(int n) {
        System.out.println("\n12. Alphabet Pattern (" + n + " rows):");
        System.out.println("----------------------------------------");
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                System.out.print((char)(65 + j) + " ");
            }
            System.out.println();
        }
    }

    // Test all patterns
    public static void main(String[] args) {
        System.out.println("============================================================");
        System.out.println("PATTERN PRINTING PROGRAMS FOR JUNIOR DEVELOPERS");
        System.out.println("============================================================");
        
        int n = 5;  // Size of patterns
        
        printSquarePattern(n);
        printRightTriangle(n);
        printInvertedRightTriangle(n);
        printPyramid(n);
        printInvertedPyramid(n);
        printDiamond(n);
        printHollowSquare(n);
        printNumberPattern(n);
        printNumberPyramid(n);
        printFloydTriangle(n);
        printPascalTriangle(n);
        printAlphabetPattern(n);
        
        System.out.println("\n============================================================");
        System.out.println("All patterns generated successfully!");
        System.out.println("============================================================");
    }
}
